Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -67,6 +67,79 @@ struct TextureMapperLayer::ComputeTransformData {
     }
 };
 
+class TextureMapperFlattenedLayer final {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    explicit TextureMapperFlattenedLayer(const IntRect& rect, double zNear, double zFar)
+        : m_rect(rect)
+        , m_zNear(zNear)
+        , m_zFar(zFar)
+    { }
+
+    inline IntRect layerRect() const
+    {
+        return m_rect;
+    }
+
+    inline bool needsUpdate() const
+    {
+        return !m_textures.size();
+    }
+
+    void update(TextureMapperPaintOptions& options, const std::function<void(TextureMapperPaintOptions&)>& paintFunction)
+    {
+        double prevZNear, prevZFar;
+        options.textureMapper.getDepthRange(prevZNear, prevZFar);
+        options.textureMapper.setDepthRange(m_zNear, m_zFar);
+
+        IntSize maxTextureSize = options.textureMapper.maxTextureSize();
+        for (int x = m_rect.x(); x < m_rect.maxX(); x += maxTextureSize.width()) {
+            for (int y = m_rect.y(); y < m_rect.maxY(); y += maxTextureSize.height()) {
+                IntRect tileRect({ x, y }, maxTextureSize);
+                tileRect.intersect(m_rect);
+
+                auto surface = options.textureMapper.acquireTextureFromPool(tileRect.size(), { BitmapTexture::Flags::SupportsAlpha });
+                {
+                    SetForScope scopedSurface(options.surface, surface.ptr());
+                    SetForScope scopedOffset(options.offset, -toIntSize(tileRect.location()));
+                    SetForScope scopedOpacity(options.opacity, 1);
+
+                    options.textureMapper.bindSurface(options.surface.get());
+                    paintFunction(options);
+
+                    surface = Ref { *options.surface };
+                }
+                m_textures.append(WTFMove(surface));
+            }
+        }
+        options.textureMapper.bindSurface(options.surface.get());
+        options.textureMapper.setDepthRange(prevZNear, prevZFar);
+    }
+
+    void paintToTextureMapper(TextureMapper& textureMapper, const FloatRect& targetRect, TransformationMatrix& modelViewMatrix, float opacity)
+    {
+        int textureIndex = 0;
+        IntSize maxTextureSize = textureMapper.maxTextureSize();
+        auto allEdgesExposed = m_rect.width() <= maxTextureSize.width() && m_rect.height() <= maxTextureSize.height() ? TextureMapper::AllEdgesExposed::Yes : TextureMapper::AllEdgesExposed::No;
+        for (int x = m_rect.x(); x < m_rect.maxX(); x += maxTextureSize.width()) {
+            for (int y = m_rect.y(); y < m_rect.maxY(); y += maxTextureSize.height()) {
+                IntRect tileRect({ x, y }, maxTextureSize);
+                tileRect.intersect(m_rect);
+
+                TransformationMatrix adjustedTransform = modelViewMatrix * TransformationMatrix::rectToRect(m_rect, targetRect);
+                textureMapper.drawTexture(*m_textures[textureIndex++].get(), tileRect, adjustedTransform, opacity, allEdgesExposed);
```

Comment:
IMHO it's dangerous to rely that the textureIndex here is valid. When any of the both two outer for loops here or in `update` are changed, this breaks. It would be better to refactor the logic into a single helper function, that centralizes the x/y paritioning <-> texture mapping logic and is shared between `update` and `paintToTextureMapper`. Note: I dislike the `update` name a bit, it's quite generic, and one doesn't really know what's happening just from reading the function name.

Reply:
How about calling it forEachTile? 😂